### PR TITLE
Problem: Changeset is not aware of new content related models

### PR DESCRIPTION
**Pull request by dkliban**
Solution: Update changesets to use Artifact, DeferredArtifact, and ContentArtifact

### CODE REVIEW

File: platform/pulpcore/app/models/content.py [hunk 1]
```diff
@@ -69,46 +104,13 @@ def __str__(self):
         return '<{} (type={}): {}>'.format(self._meta.object_name, cast.TYPE, natural_key_string)
 
 
-class Artifact(Model):
+class ContentArtifact(Model):
```
Comment (jortel) [inline platform/pulpcore/app/models/content.py]: Need a unique index on (content, artifact, relative_path).

File: platform/pulpcore/app/models/content.py [hunk 2]
```diff
@@ -3,10 +3,44 @@
...
+    file = models.FileField(db_index=True, upload_to=storage_path, max_length=255)
+    downloaded = models.BooleanField(db_index=True, default=False)
+    requested = models.BooleanField(db_index=True, default=False)
+    size = models.IntegerField(blank=True, null=True)
+    md5 = models.CharField(max_length=32, blank=True, null=True)
+    sha1 = models.CharField(max_length=40, blank=True, null=True)
+    sha224 = models.CharField(max_length=56, blank=True, null=True)
+    sha256 = models.CharField(max_length=64, blank=True, null=True)
+    sha384 = models.CharField(max_length=96, blank=True, null=True)
+    sha512 = models.CharField(max_length=128, blank=True, null=True)
```
Comment (jortel) [inline platform/pulpcore/app/models/content.py]: I assume an artifact cannot be created without at least 1 digest, correct?
And, the query to find an existing artifact is by matching one of the digests?

like:
```
where md5 = <md5> or,
      sha1 = <sha1> or,
      sha224 = <sha224> or,
      sha256 = <sha256> or,
      ...
```
only for non-null digests.

If so, won't we need indexes on all of the digests?

**Comment by jortel** [comment]
The FK in `DownloadCatalog` to Artifact will need to be changed to ContentArtifact.

File: platform/pulpcore/app/models/content.py [hunk 3]
```diff
@@ -63,11 +92,33 @@ def storage_path(self, name):
...
-    sha224 = models.CharField(max_length=56, blank=False, null=False, unique=False, db_index=True)
-    sha256 = models.CharField(max_length=64, blank=False, null=False, unique=True, db_index=True)
-    sha384 = models.CharField(max_length=96, blank=False, null=False, unique=True, db_index=True)
-    sha512 = models.CharField(max_length=128, blank=False, null=False, unique=True, db_index=True)
+    artifact = models.ForeignKey(Artifact, on_delete=models.CASCADE, null=True)
+    content = models.ForeignKey(Content, on_delete=models.CASCADE)
+    relative_path = models.CharField(max_length=64)
+
+
+class DeferredArtifact(Model):
```
Comment (jortel) [inline platform/pulpcore/app/models/content.py]: Suggest another name.  Perhaps `PendingArtifact` might be better since they are created even when downloading is not deferred.

File: platform/pulpcore/app/models/content.py [hunk 4]
```diff
@@ -1,12 +1,49 @@
...
+    """
+    def storage_path(self, name):
+        """
+        Callable used by FileField to determine where the uploaded file should be stored.
+
+        Args:
+            name (str): Original name of uploaded file. It is ignored by this method because the
+                sha256 checksum is used to determine a file path instead.
+        """
+        return default_storage.get_artifact_path(self.sha256)
```
Comment (jortel) [inline platform/pulpcore/app/models/content.py]: It has been my experience that for this to work, the Artifact must be a `deconstructable`.  Otherwise, it just fails silently.  Did you validate the file is actually written to the storage location?

Also, I thought we agreed a cleaner approach would be to add a path calculator to the storage module and add as an attribute to the `FileStorage`.  See:  https://github.com/pulp/pulp/pull/3100.

File: platform/pulpcore/download/writer.py [hunk 5]
```diff
@@ -92,6 +93,9 @@ def __init__(self, path):
         """
         super(FileWriter, self).__init__()
         self.path = path
+        self.hashers = {}
+        for hasher in hashlib.algorithms_guaranteed:
+            self.hashers[hasher] = getattr(hashlib, hasher)()
```
Comment (jortel) [inline platform/pulpcore/download/writer.py]: For better separation of concerns, a `Writer` should not be generating digests.  This is also redundant with digests being calculated by `Validation`.  There are cleaner ways to hook this in.  For example: adding a buffer event listener to calculate all the digests to be used when creating an `Artifact`.

File: platform/pulpcore/app/viewsets/content.py [hunk 6]
```diff
@@ -44,6 +45,28 @@ class ContentViewSet(NamedModelViewSet):
     serializer_class = ContentSerializer
     filter_class = ContentFilter
 
+    @transaction.atomic
+    def create(self, request):
```
Comment (jortel) [inline platform/pulpcore/app/viewsets/content.py]: I haven't seen this kind of _helper_ method on other models.  So far, we've made a deliberate effort to keep the Pulp domain model and the data model separated.  It seems like this functionality belongs somewhere else.

File: plugin/pulpcore/plugin/changeset/iterator.py [hunk 7]
```diff
@@ -166,13 +169,21 @@ def _iter(self):
         Yields:
             Download: The appropriate download object.
         """
-        for artifact in ArtifactIterator(self.content):
-            if self.deferred or artifact.model.downloaded:
+        for remote_artifact in ArtifactIterator(self.content):
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/iterator.py]: The variable named `artifact` refers to a `RemoteArtifact` consistently in the code base, specifically naming this `remote_artifact` introduces confusion.

File: plugin/pulpcore/plugin/changeset/iterator.py [hunk 8]
```diff
@@ -199,15 +210,15 @@ def _iter(self):
         does not have any un-downloaded artifacts, A NopArtifact is yielded.
 
         Yields:
-            RemoteArtifact: The flattened artifacts.
+            DelayedArtifact: The flattened artifacts.
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/iterator.py]: I don't think you mean `DelayedArtifact` here.

File: plugin/pulpcore/plugin/changeset/model.py [hunk 9]
```diff
@@ -131,13 +135,42 @@ def save(self):
...
+                for remote_artifact in self.artifacts:
+                    if not remote_artifact.content_artifact.artifact_id:
+                        if not deferred:
+                            artifact = self.validate_and_create_artifact(remote_artifact.model,
+                                                                         remote_artifact.download)
+                            remote_artifact.content_artifact.artifact = artifact
+                        remote_artifact.model.save()
+                        remote_artifact.content_artifact.save()
+
+    def validate_and_create_artifact(self, remote_artifact, download):
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/model.py]: Each download is configured with digest validation.  If the validation fails, the download fails.  We don't need/want to do it here as well.

File: plugin/pulpcore/plugin/changeset/model.py [hunk 10]
```diff
@@ -147,41 +180,47 @@ class RemoteArtifact(Remote):
...
-        >>> from pulpcore.plugin.models import Artifact
+        >>> from pulpcore.plugin.models import ContentArtifact, DeferredArtifact
         >>>
-        >>> model = Artifact(...)  # DB model instance.
+        >>> model = DeferredArtifact(...)
+        >>> content_artifact = ContentArtifact(...)
         >>> download = ...
         >>> ...
-        >>> artifact = RemoteArtifact(model, download)
+        >>> artifact = RemoteArtifact(model, content_artifact, download)
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/model.py]: Seems like we can just add relative_path to `RemoteArtifact` and simplify the interface by not having the plugin writer deal with it specifically.

File: plugin/pulpcore/plugin/changeset/model.py [hunk 11]
```diff
@@ -147,41 +180,47 @@ class RemoteArtifact(Remote):
...
         >>>
     """
 
     __slots__ = (
         'content',
+        'content_artifact',
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/model.py]: Let's make this `relative_path` and manage the `ContentArtifact` for the plugin writer.

File: plugin/pulpcore/plugin/download/factory.py [hunk 12]
```diff
@@ -106,8 +108,9 @@ def _https(self, url, path=None, artifact=None):
         Returns:
             HttpDownload: An https download.
         """
-        if artifact:
-            _path = artifact.relative_path
+        if remote_artifact:
+            import os
+            _path = os.path.join("/var/lib/pulp/tmp", remote_artifact.sha256)
```
Comment (jortel) [inline plugin/pulpcore/plugin/download/factory.py]: Why are we doing this?  We should not be hard coding paths.  the downloader should be configured to download into the working_dir which is the CWD.

File: plugin/pulpcore/plugin/download/factory.py [hunk 13]
```diff
@@ -118,7 +121,6 @@ def _https(self, url, path=None, artifact=None):
         download.user.name = self.importer.basic_auth_user
         download.user.password = self.importer.basic_auth_password
         download.proxy_url = self.importer.proxy_url
-        self._add_validation(download, artifact)
```
Comment (jortel) [inline plugin/pulpcore/plugin/download/factory.py]: Why is the call to add validation removed?

File: plugin/pulpcore/plugin/changeset/iterator.py [hunk 14]
```diff
@@ -166,13 +169,21 @@ def _iter(self):
...
             Download: The appropriate download object.
         """
-        for artifact in ArtifactIterator(self.content):
-            if self.deferred or artifact.model.downloaded:
+        for remote_artifact in ArtifactIterator(self.content):
+            q = Q()
+            for algorithm in hashlib.algorithms_guaranteed:
+                if getattr(remote_artifact.model, algorithm):
+                    kwargs = {algorithm: getattr(remote_artifact.model, algorithm)}
+                    q |= Q(**kwargs)
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/iterator.py]: Django already has a convention for getting the natural key for a model.  The `natural_key()` method normally returns a tuple which is obviously not sufficient for this.  But, I think we should follow the same pattern and have a method on `RemoteArtifact` that can be used to fetch it from the DB.  Instead of here.

File: plugin/pulpcore/plugin/changeset/iterator.py [hunk 15]
```diff
@@ -166,13 +169,21 @@ def _iter(self):
...
+            try:
+                artifact = Artifact.objects.get(q)
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/iterator.py]: This query needs to be done in batches.  For performance, we cannot be hitting the DB for every artifact.  Other code in the _changeset_ package does this.

File: plugin/pulpcore/plugin/changeset/iterator.py [hunk 16]
```diff
@@ -166,13 +169,21 @@ def _iter(self):
...
+            except Artifact.DoesNotExist:
+                artifact = None
+            if artifact or self.deferred:
                 download = NopDownload()
-                download.attachment = artifact
-                artifact.path = None
+                download.attachment = remote_artifact
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/iterator.py]: Setting the `Artifact.path` is how it's know that a file was actually downloaded for the artifact.  This should not change.

File: plugin/pulpcore/plugin/changeset/iterator.py [hunk 17]
```diff
@@ -166,13 +169,21 @@ def _iter(self):
...
             else:
-                download = artifact.download
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/iterator.py]: As mentioned earlier.  The convention in the package is that:

`content` is `RemoteContent`
and
`artifact` is `RemoteArtifact`

And that `model` was used for DB models.  If you want to change this convention, you need to do a mass rename.  Though it seems outside the scope of this work.

File: plugin/pulpcore/plugin/changeset/iterator.py (hunk 8 above)
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/iterator.py]: This should be `RemoteArtifact`.

File: plugin/pulpcore/plugin/changeset/model.py [hunk 18]
```diff
@@ -147,41 +180,47 @@ class RemoteArtifact(Remote):
...
         'download',
         'path'
     )
 
-    def __init__(self, model, download):
+    def __init__(self, model, content_artifact, download):
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/model.py]: content_artifact removed.

File: plugin/pulpcore/plugin/changeset/model.py [hunk 19]
```diff
@@ -91,17 +95,17 @@ def update(self, model):
             model (pulpcore.plugin.Content): A fetched content model object.
         """
         self.model = model
-        known = {a.model.relative_path: a for a in self.artifacts}
+        known = {a.content_artifact.relative_path: a for a in self.artifacts}
         self.artifacts.clear()
-        for artifact in model.artifacts.all():
+        for content_artifact in ContentArtifact.objects.filter(content=model):
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/model.py]: Unfortunate the django does not support query through the explicit thru table :/

File: plugin/pulpcore/plugin/changeset/model.py [hunk 20]
```diff
@@ -131,13 +135,42 @@ def save(self):
...
+            data['size'] = remote_artifact.size
+        for hasher in hashlib.algorithms_guaranteed:
+            if getattr(remote_artifact, hasher, None):
+                data[hasher] = getattr(remote_artifact, hasher, None)
+
+        serializer_class = get_plugin_config("pulp_app").named_serializers["ArtifactSerializer"]
+        serializer = serializer_class(data=data)
+        serializer.validate(data=data)
+        serializer.is_valid(raise_exception=True)
+        return serializer.save()
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/model.py]: Seems this logic should be moved to the `RemoteArtifact.save()`.

File: plugin/pulpcore/plugin/changeset/model.py [hunk 21]
```diff
@@ -131,13 +135,42 @@ def save(self):
...
+                    if not remote_artifact.content_artifact.artifact_id:
+                        if not deferred:
+                            artifact = self.validate_and_create_artifact(remote_artifact.model,
+                                                                         remote_artifact.download)
+                            remote_artifact.content_artifact.artifact = artifact
+                        remote_artifact.model.save()
+                        remote_artifact.content_artifact.save()
+
+    def validate_and_create_artifact(self, remote_artifact, download):
+        """
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/model.py]: I don't know what decisions were made with the epic _validation_ discussion on pulp-dev but seem wrong for platform _core_ components to be using the serializer to do validation and creation.  I don't think we do this anywhere else.

File: plugin/pulpcore/plugin/changeset/model.py [hunk 22]
```diff
@@ -131,13 +135,42 @@ def save(self):
...
                 is_duplicate = True
             else:
-                for artifact in self.artifacts:
-                    artifact.model.content = self.model
-                    artifact.save()
-        if is_duplicate:
-            model = type(self.model)
-            content = model.objects.get(**self.key)
-            self.update(content)
+
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/model.py]: extra blank line.

File: plugin/pulpcore/plugin/changeset/model.py [hunk 23]
```diff
@@ -131,13 +135,42 @@ def save(self):
...
+                for remote_artifact in self.artifacts:
+                    if not remote_artifact.content_artifact.artifact_id:
+                        if not deferred:
```
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/model.py]: The decision to create the artifact should be based on decisions made by the `ArtifactIterator` already.  When the `RemoteArtifact.path` is set, the file has been downloaded and real `Artifact` should be created.

File: plugin/pulpcore/plugin/changeset/model.py (hunk 20 above)
Comment (jortel) [inline plugin/pulpcore/plugin/changeset/model.py]: What about race conditions whereby the artifact does not exist when the artifact iterator checks but another thread has created it before this code tries to insert it.  The dup needs to be handled.

**Comment by mhrivnak** [comment]
It sounds like this is getting replaced by this new work: https://pulp.plan.io/issues/2942

To avoid confusion (which confused me just now), I'm closing this PR. Obviously please re-open if I'm misinterpreting.
